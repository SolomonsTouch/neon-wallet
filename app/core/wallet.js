// @flow
import { ASSETS, TOKENS } from './constants'
import { wallet } from 'neon-js'

const MIN_PASSPHRASE_LEN = 4

export const validatePassphraseLength = (passphrase: string): boolean => passphrase.length >= MIN_PASSPHRASE_LEN

export const isToken = (symbol: SymbolType) => Object.keys(TOKENS).includes(symbol)

export const obtainBalance = (balances: Object, symbol: SymbolType) => {
  return balances[symbol] || 0
}

export const validateTransactionBeforeSending = (balance: number, sendEntry: SendEntryType) => {
  const { address, amount, symbol } = sendEntry

  if (!address || !amount) {
    return 'Please specify an address and amount.'
  }

  if (symbol !== ASSETS.NEO && symbol !== ASSETS.GAS && !isToken(symbol)) {
    return 'That asset is not NEO, GAS or NEP-5 Token.'
  }

  try {
    if (wallet.isAddress(address) !== true || address.charAt(0) !== 'A') {
      return 'The address you entered was not valid.'
    }
  } catch (e) {
    return 'The address you entered was not valid.'
  }

  if (symbol === ASSETS.NEO && parseFloat(amount) !== parseInt(amount)) { // check for fractional NEO
    return 'You cannot send fractional amounts of NEO.'
  }

  if (parseFloat(amount) > balance) {
    return `You do not have enough ${symbol} to send.`
  }

  if (parseFloat(amount) <= 0) { // check for negative/zero asset
    return 'You cannot send zero or negative amounts of an asset.'
  }

  return null
}

export const validateTransactionsBeforeSending = (balances: Object, sendEntries: Array<SendEntryType>) => {
  const getValidationError = (sendEntry) => {
    const balance = obtainBalance(balances, sendEntry.symbol)
    return validateTransactionBeforeSending(balance, sendEntry)
  }

  const errorEntry = sendEntries.find(getValidationError)

  return errorEntry ? getValidationError(errorEntry) : null
}
