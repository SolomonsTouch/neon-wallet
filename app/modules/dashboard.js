// @flow
import { LOGOUT } from './account'

const initialState = {}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
