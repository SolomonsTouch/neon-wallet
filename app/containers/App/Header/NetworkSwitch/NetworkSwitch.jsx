// @flow
import React, { Component } from 'react'

import { NETWORK } from '../../../../core/constants'

import headerStyles from '../Header.scss'

type Props = {
  net: NetworkType,
  setNetwork: Function,
  loadWalletData: Function
}

export default class NetworkSwitch extends Component<Props> {
  chooseNetwork = (e: Object) => {
    const { setNetwork, loadWalletData } = this.props
    const newNet = e.target.value
    setNetwork(newNet)
    loadWalletData(false)
  }

  render () {
    const { net } = this.props
    return (
      <div id='network' className={headerStyles.navBarItem}>
        <span className={headerStyles.navBarItemLabel}>Running on</span>
        <select defaultValue={net} onChange={this.chooseNetwork} className='networkSelector'>
          <option value={NETWORK.MAIN}>{NETWORK.MAIN}</option>
          <option value={NETWORK.TEST}>{NETWORK.TEST}</option>
        </select>
      </div>
    )
  }
}
