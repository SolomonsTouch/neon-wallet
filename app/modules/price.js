// @flow
import { api } from 'neon-js'

import asyncWrap from '../core/asyncHelper'
import { DEFAULT_CURRENCY_CODE } from '../core/constants'

// Constants
export const SET_NEO_PRICE = 'SET_NEO_PRICE'
export const SET_GAS_PRICE = 'SET_GAS_PRICE'
export const RESET_PRICE = 'RESET_PRICE'
export const SET_CURRENCY = 'SET_CURRENCY'

// Actions
export function setNEOPrice (NEO: string) {
  return {
    type: SET_NEO_PRICE,
    payload: { NEO }
  }
}

export function setGASPrice (GAS: string) {
  return {
    type: SET_GAS_PRICE,
    payload: { GAS }
  }
}

export function setCurrency (currency: string) {
  return {
    type: SET_CURRENCY,
    payload: { currency }
  }
}

export function resetPrice () {
  return {
    type: RESET_PRICE
  }
}

export const getMarketPriceUSD = () => async (dispatch: DispatchType, getState: GetStateType) => {
  // If API dies, still display balance - ignore _err
  const [_err, price] = await asyncWrap(api.cmc.getPrice('NEO', getCurrency(getState()))) // eslint-disable-line
  return dispatch(setNEOPrice(price))
}

export const getGasMarketPriceUSD = () => async (dispatch: DispatchType, getState: GetStateType) => {
  // If API dies, still display balance - ignore _err
  const [_err, price] = await asyncWrap(api.cmc.getPrice('GAS', getCurrency(getState()))) // eslint-disable-line
  return dispatch(setGASPrice(price))
}

// state getters
export const getNEOPrice = (state: Object) => state.price.NEO
export const getGASPrice = (state: Object) => state.price.GAS
export const getCurrency = (state: Object) => state.price.currency

const initialState = {
  NEO: 0,
  GAS: 0,
  currency: DEFAULT_CURRENCY_CODE
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case SET_NEO_PRICE:
      const { NEO } = action.payload
      return {
        ...state,
        NEO
      }
    case SET_GAS_PRICE:
      const { GAS } = action.payload
      return {
        ...state,
        GAS
      }
    case SET_CURRENCY:
      const { currency } = action.payload
      return {
        ...state,
        NEO: 0,
        GAS: 0,
        currency
      }
    case RESET_PRICE:
      return {
        ...state,
        NEO: 0,
        GAS: 0
      }
    default:
      return state
  }
}
