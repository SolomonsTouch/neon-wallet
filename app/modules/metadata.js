// @flow
import axios from 'axios'
import { api } from 'neon-js'
import storage from 'electron-json-storage'

import { showWarningNotification } from './notifications'
import { setCurrency } from './price'

import { NETWORK, EXPLORERS, NEON_WALLET_RELEASE_LINK, NOTIFICATION_POSITIONS } from '../core/constants'
import asyncWrap from '../core/asyncHelper'

import { version } from '../../package.json'

// Constants
export const SET_HEIGHT = 'SET_HEIGHT'
export const SET_NETWORK = 'SET_NETWORK'
export const SET_EXPLORER = 'SET_EXPLORER'

// Actions
export function setNetwork (net: NetworkType) {
  const network = net === NETWORK.MAIN ? NETWORK.MAIN : NETWORK.TEST
  return {
    type: SET_NETWORK,
    payload: { network }
  }
}

export function setBlockHeight (blockHeight: number) {
  return {
    type: SET_HEIGHT,
    payload: { blockHeight }
  }
}

export function setBlockExplorer (blockExplorer: ExplorerType) {
  return {
    type: SET_EXPLORER,
    payload: { blockExplorer }
  }
}

export const checkVersion = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const net = getNetwork(state)
  const apiEndpoint = api.neonDB.getAPIEndpoint(net)

  const [err, res] = await asyncWrap(axios.get(`${apiEndpoint}/v2/version`))
  const shouldUpdate = res && res.data && res.data.version !== version
  if (err || shouldUpdate) {
    const link = `<a href='${NEON_WALLET_RELEASE_LINK}' target='_blank' class="notification-link">${NEON_WALLET_RELEASE_LINK}</a>`
    const message = err ? `Error checking wallet version! Please make sure you have downloaded the latest version: ${link}`
      : `Your wallet is out of date! Please download the latest version from ${link}`
    return dispatch(showWarningNotification({
      message,
      autoDismiss: 0,
      stack: true,
      position: NOTIFICATION_POSITIONS.BOTTOM_CENTER
    }))
  }
}

export const initSettings = () => async (dispatch: DispatchType) => {
  // eslint-disable-next-line
  storage.get('settings', (error, settings) => {
    if (settings.blockExplorer !== null && settings.blockExplorer !== undefined) {
      dispatch(setBlockExplorer(settings.blockExplorer))
    }

    if (settings.currency !== null && settings.currency !== undefined) {
      dispatch(setCurrency(settings.currency))
    }
  })
}

export const syncBlockHeight = (net: NetworkType) => async (dispatch: DispatchType) => {
  const [_err, blockHeight] = await asyncWrap(api.neonDB.getWalletDBHeight(net)) // eslint-disable-line
  return dispatch(setBlockHeight(blockHeight))
}

// state getters
export const getBlockHeight = (state: Object) => state.metadata.blockHeight
export const getNetwork = (state: Object) => state.metadata.network
export const getBlockExplorer = (state: Object) => state.metadata.blockExplorer

const initialState = {
  blockHeight: 0,
  network: NETWORK.MAIN,
  blockExplorer: EXPLORERS.NEO_TRACKER
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case SET_HEIGHT:
      const { blockHeight } = action.payload
      return {
        ...state,
        blockHeight
      }
    case SET_EXPLORER:
      const { blockExplorer } = action.payload
      return {
        ...state,
        blockExplorer
      }
    case SET_NETWORK:
      const { network } = action.payload
      return {
        ...state,
        network
      }
    default:
      return state
  }
}
