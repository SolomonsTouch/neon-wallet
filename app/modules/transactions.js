// @flow
/* eslint-disable camelcase */
import { api, sc, u, wallet } from 'neon-js'
import { flatMap } from 'lodash'

import {
  setTransactionHistory,
  getBalances,
  getScriptHashForNetwork
} from './wallet'
import {
  showErrorNotification,
  showInfoNotification,
  showSuccessNotification
} from './notifications'
import {
  getWIF,
  getPublicKey,
  getSigningFunction,
  getAddress,
  LOGOUT,
  getIsHardwareLogin
} from './account'
import { getNetwork } from './metadata'

import { isToken, validateTransactionsBeforeSending } from '../core/wallet'
import { ASSETS } from '../core/constants'
import { adjustDecimalAmountForTokenTransfer } from '../core/nep5'
import asyncWrap from '../core/asyncHelper'
import { toNumber, toBigNumber } from '../core/math'
import { toFixedDecimals, COIN_DECIMAL_LENGTH } from '../core/formatters'

import { log } from '../util/Logs'

// Constants
export const LOADING_TRANSACTIONS = 'LOADING_TRANSACTIONS'

export const setIsLoadingTransaction = (isLoading: boolean) => ({
  type: LOADING_TRANSACTIONS,
  payload: {
    isLoadingTransactions: isLoading
  }
})

export const syncTransactionHistory = (
  net: NetworkType,
  address: string
) => async (dispatch: DispatchType) => {
  dispatch(setIsLoadingTransaction(true))
  const [err, transactions] = await asyncWrap(
    api.neonDB.getTransactionHistory(net, address)
  )
  if (!err && transactions) {
    const txs = transactions.map(
      ({ NEO, GAS, txid, block_index }: TransactionHistoryType) => ({
        txid,
        [ASSETS.NEO]: toFixedDecimals(NEO, 0),
        [ASSETS.GAS]: toBigNumber(GAS).round(COIN_DECIMAL_LENGTH).toString()
      })
    )
    dispatch(setIsLoadingTransaction(false))
    dispatch(setTransactionHistory(txs))
  } else {
    dispatch(setIsLoadingTransaction(false))
  }
}

const extractTokens = (sendEntries: Array<SendEntryType>) => {
  return sendEntries.filter(({ symbol }) => isToken(symbol))
}

const extractAssets = (sendEntries: Array<SendEntryType>) => {
  return sendEntries.filter(({ symbol }) => !isToken(symbol))
}

const buildIntents = (sendEntries: Array<SendEntryType>) => {
  const assetEntries = extractAssets(sendEntries)
  // $FlowFixMe
  return flatMap(assetEntries, ({ address, amount, symbol }) =>
    api.makeIntent(
      {
        [symbol]: toNumber(amount)
      },
      address
    )
  )
}

const buildIntentsForInvocation = (
  sendEntries: Array<SendEntryType>,
  fromAddress: string
) => {
  const intents = buildIntents(sendEntries)

  if (intents.length > 0) {
    return intents
  } else {
    return buildIntents([
      {
        address: fromAddress,
        amount: '0.00000001',
        symbol: ASSETS.GAS
      }
    ])
  }
}

const buildTransferScript = (
  net: NetworkType,
  sendEntries: Array<SendEntryType>,
  fromAddress: string
) => {
  const tokenEntries = extractTokens(sendEntries)
  const fromAcct = new wallet.Account(fromAddress)
  const scriptBuilder = new sc.ScriptBuilder()

  tokenEntries.forEach(({ address, amount, symbol }) => {
    const toAcct = new wallet.Account(address)
    const scriptHash = getScriptHashForNetwork(net, symbol)
    const args = [
      u.reverseHex(fromAcct.scriptHash),
      u.reverseHex(toAcct.scriptHash),
      adjustDecimalAmountForTokenTransfer(amount)
    ]

    scriptBuilder.emitAppCall(scriptHash, 'transfer', args)
  })

  return scriptBuilder.str
}

const makeRequest = (sendEntries: Array<SendEntryType>, config: Object) => {
  const script = buildTransferScript(config.net, sendEntries, config.address)

  if (script === '') {
    return api.sendAsset({ ...config, intents: buildIntents(sendEntries) })
  } else {
    return api.doInvoke({
      ...config,
      intents: buildIntentsForInvocation(sendEntries, config.address),
      script,
      gas: 0
    })
  }
}

export const sendTransaction = (sendEntries: Array<SendEntryType>) => async (
  dispatch: DispatchType,
  getState: GetStateType
): Promise<*> => {
  const state = getState()
  const wif = getWIF(state)
  const fromAddress = getAddress(state)
  const net = getNetwork(state)
  const balances = getBalances(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)
  const isHardwareSend = getIsHardwareLogin(state)

  const rejectTransaction = (message: string) =>
    dispatch(showErrorNotification({ message }))

  const error = validateTransactionsBeforeSending(balances, sendEntries)

  if (error) {
    return rejectTransaction(error)
  }

  dispatch(
    showInfoNotification({ message: 'Sending Transaction...', autoDismiss: 0 })
  )

  log(
    net,
    'SEND',
    fromAddress,
    // $FlowFixMe
    sendEntries.map(({ address, amount, symbol }) => ({
      to: address,
      asset: symbol,
      amount: parseFloat(amount)
    }))
  )

  if (isHardwareSend) {
    dispatch(
      showInfoNotification({
        message: 'Please sign the transaction on your hardware device',
        autoDismiss: 0
      })
    )
  }

  const [err, config] = await asyncWrap(
    makeRequest(sendEntries, {
      net,
      address: fromAddress,
      publicKey,
      privateKey: new wallet.Account(wif).privateKey,
      signingFunction: isHardwareSend ? signingFunction : null
    })
  )

  if (err || !config || !config.response || !config.response.result) {
    console.log(err)
    return rejectTransaction('Transaction failed!')
  } else {
    return dispatch(
      showSuccessNotification({
        message:
          'Transaction complete! Your balance will automatically update when the blockchain has processed it.'
      })
    )
  }
}

// state getters
export const getIsLoadingTransactions = (state: Object) =>
  state.transactions.isLoadingTransactions

const initialState = {
  isLoadingTransactions: false
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case LOADING_TRANSACTIONS:
      const { isLoadingTransactions } = action.payload
      return {
        ...state,
        isLoadingTransactions
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
