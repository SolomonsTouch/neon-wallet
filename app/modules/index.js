// @flow
import { combineReducers } from 'redux'
import account from './account'
import generateWallet from './generateWallet'
import transactions from './transactions'
import metadata from './metadata'
import wallet from './wallet'
import claim from './claim'
import dashboard from './dashboard'
import notifications from './notifications'
import modal from './modal'
import price from './price'
import addressBook from './addressBook'

export default combineReducers({
  account,
  generateWallet,
  wallet,
  transactions,
  dashboard,
  metadata,
  claim,
  notifications,
  modal,
  price,
  addressBook
})
