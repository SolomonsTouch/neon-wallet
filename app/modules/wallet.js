// @flow
import { api } from 'neon-js'
import { mapValues } from 'lodash'

import { syncTransactionHistory } from './transactions'
import { syncAvailableClaim } from './claim'
import { syncBlockHeight, getNetwork } from './metadata'
import { LOGOUT, getAddress } from './account'
import { getMarketPriceUSD, getGasMarketPriceUSD } from './price'
import { showErrorNotification } from './notifications'

import { ASSETS, TOKENS, TOKENS_TEST, NETWORK } from '../core/constants'
import asyncWrap from '../core/asyncHelper'
import { COIN_DECIMAL_LENGTH } from '../core/formatters'
import { toBigNumber } from '../core/math'

const TOKEN_PAIRS = Object.entries(TOKENS)

export const getScriptHashForNetwork = (net: NetworkType, symbol: any) => {
  if (net === NETWORK.TEST && TOKENS_TEST[symbol]) {
    return TOKENS_TEST[symbol]
  }
  return TOKENS[symbol]
}

// Constants
export const SET_BALANCE = 'SET_BALANCE'
export const SET_NEO_PRICE = 'SET_NEO_PRICE'
export const SET_GAS_PRICE = 'SET_GAS_PRICE'
export const RESET_PRICES = 'RESET_PRICES'
export const SET_TRANSACTION_HISTORY = 'SET_TRANSACTION_HISTORY'
export const SET_TOKENS = 'SET_TOKENS'
export const SET_IS_LOADED = 'SET_IS_LOADED'

export const setIsLoaded = (loaded: boolean) => ({
  type: SET_IS_LOADED,
  payload: {
    loaded
  }
})

// Actions
export function setBalance (NEO: string, GAS: string) {
  return {
    type: SET_BALANCE,
    payload: { NEO, GAS }
  }
}

export function setTransactionHistory (transactions: Array<Object>) {
  return {
    type: SET_TRANSACTION_HISTORY,
    payload: { transactions }
  }
}

export function setTokens (tokens: Object) {
  return {
    type: SET_TOKENS,
    payload: { tokens }
  }
}

export const retrieveBalance = (net: NetworkType, address: string) => async (
  dispatch: DispatchType
) => {
  // If API dies, still display balance - ignore _err
  const [_err, resultBalance] = await asyncWrap(
    api.neonDB.getBalance(net, address)
  ) // eslint-disable-line
  if (_err) {
    return dispatch(
      showErrorNotification({
        message: `Could not retrieve NEO/GAS balance`,
        stack: true
      })
    )
  } else {
    return dispatch(
      setBalance(
        String(resultBalance.NEO.balance),
        toBigNumber(resultBalance.GAS.balance)
          .round(COIN_DECIMAL_LENGTH)
          .toString()
      )
    )
  }
}

export const loadWalletData = (silent: boolean = true) => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const net = getNetwork(state)
  const address = getAddress(state)

  if (!silent) {
    dispatch(setIsLoaded(false))
  }
  dispatch(syncTransactionHistory(net, address))
  dispatch(syncAvailableClaim(net, address))
  dispatch(syncBlockHeight(net))
  dispatch(getMarketPriceUSD())
  dispatch(getGasMarketPriceUSD())
  await Promise.all([
    dispatch(retrieveTokensBalance()),
    dispatch(retrieveBalance(net, address))
  ])
  dispatch(setIsLoaded(true))
  return true
}

export const retrieveTokensBalance = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const net = getNetwork(state)
  const address = getAddress(state)

  const tokens = getInitialTokenBalance()
  for (let [symbol] of TOKEN_PAIRS) {
    const scriptHash = getScriptHashForNetwork(net, symbol)
    // override scripthash with test if on test net
    // eslint-disable-next-line
    const [_error, tokenRpcEndpoint] = await asyncWrap(
      api.neonDB.getRPCEndpoint(net)
    )
    const [err, tokenResults] = await asyncWrap(
      api.nep5.getToken(tokenRpcEndpoint, scriptHash, address)
    )

    if (!err) {
      tokens[symbol] = {
        ...tokenResults,
        balance:
          tokenResults.balance === null
            ? '0'
            : toBigNumber(tokenResults.balance)
              .round(COIN_DECIMAL_LENGTH)
              .toString(),
        scriptHash
      }
    } else {
      dispatch(
        showErrorNotification({
          message: `could not retrieve ${symbol} balance`,
          stack: true
        })
      )
    }
  }

  return dispatch(setTokens(tokens))
}

// state getters
export const getNEO = (state: Object): string => state.wallet.NEO
export const getGAS = (state: Object): string => state.wallet.GAS
export const getTransactions = (state: Object) => state.wallet.transactions
export const getTokens = (state: Object) => state.wallet.tokens
export const getIsLoaded = (state: Object) => state.wallet.loaded

export const getBalances = (state: Object) => ({
  [ASSETS.NEO]: getNEO(state),
  [ASSETS.GAS]: getGAS(state),
  ...mapValues(getTokens(state), token => token.balance)
})

const getInitialTokenBalance = () => {
  const tokens = {}
  Object.keys(TOKENS).forEach(symbol => {
    tokens[symbol] = {
      symbol,
      scriptHash: TOKENS[symbol],
      balance: '0'
    }
  })
  return tokens
}

type State = {
  loaded: boolean,
  NEO: string,
  GAS: string,
  transactions: Array<TransactionHistoryType>,
  tokens: Object
}

const initialState = {
  loaded: false,
  NEO: '0',
  GAS: '0',
  transactions: [],
  tokens: getInitialTokenBalance()
}

export default (state: State = initialState, action: ReduxAction) => {
  switch (action.type) {
    case SET_BALANCE:
      const { NEO, GAS } = action.payload
      return {
        ...state,
        NEO,
        GAS
      }
    case SET_TRANSACTION_HISTORY:
      const { transactions } = action.payload
      return {
        ...state,
        transactions
      }
    case SET_TOKENS:
      const { tokens } = action.payload
      return {
        ...state,
        tokens
      }
    case SET_IS_LOADED:
      const { loaded } = action.payload
      return {
        ...state,
        loaded
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
