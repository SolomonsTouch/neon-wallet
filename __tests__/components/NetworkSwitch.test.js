import React from 'react'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { shallow } from 'enzyme'

import { SET_HEIGHT, SET_NETWORK } from '../../app/modules/metadata'
import { SET_BALANCE, SET_TRANSACTION_HISTORY, SET_IS_LOADED } from '../../app/modules/wallet'
import { LOADING_TRANSACTIONS } from '../../app/modules/transactions'
import NetworkSwitch from '../../app/containers/App/Header/NetworkSwitch'

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { version } from '../../package.json'

const axiosMock = new MockAdapter(axios)
axiosMock
  .onGet('http://testnet-api.wallet.cityofzion.io/v2/version')
  .reply(200, { version })
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/NEO/', { params: { convert: 'USD' } })
  .reply(200, [ { price_usd: 24.50 } ])
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/GAS/', { params: { convert: 'USD' } })
  .reply(200, [ { price_usd: 18.20 } ])
jest.mock('neon-js')

const setup = () => {
  const props = {
    net: 'MainNet',
    setNetwork: jest.fn(),
    loadWalletData: jest.fn()
  }
  const wrapper = shallow(<NetworkSwitch {...props} />)

  return {
    wrapper
  }
}

describe('NetworkSwitch', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })

  test('correctly renders MainNet initially', () => {
    const { wrapper } = setup()

    const networkSelectorElement = wrapper.find('.networkSelector').getElement()

    expect(networkSelectorElement.props.defaultValue).toEqual('MainNet')
  })

  test('switches to the correct network when chosen from the dropdown', async () => {
    const { wrapper } = setup()

    const instance = wrapper.instance()
    const networkSelector = wrapper.find('.networkSelector')
    networkSelector.simulate('change', { target: { value: 'TestNet' } })

    expect(instance.props.setNetwork).toHaveBeenCalledWith('TestNet')
    expect(instance.props.loadWalletData).toHaveBeenCalled()

    networkSelector.simulate('change', { target: { value: 'MainNet' } })

    expect(instance.props.setNetwork).toHaveBeenCalledWith('MainNet')
    expect(instance.props.loadWalletData).toHaveBeenCalled()
  })
})
