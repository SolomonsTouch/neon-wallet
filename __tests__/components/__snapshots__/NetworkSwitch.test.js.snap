// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkSwitch renders without crashing 1`] = `
<div
  className="navBarItem"
  id="network"
>
  <span
    className="navBarItemLabel"
  >
    Running on
  </span>
  <select
    className="networkSelector"
    defaultValue="MainNet"
    onChange={[Function]}
  >
    <option
      value="MainNet"
    >
      MainNet
    </option>
    <option
      value="TestNet"
    >
      TestNet
    </option>
  </select>
</div>
`;
